### ISSUE TITLE
Review use of YAML

### ISSUE THREAD

**Issue by scottdurow** [microsoft/PowerApps-Language-Tooling#111]
YAML is like [marmite ](https://en.wikipedia.org/wiki/Marmite) - you either love it or hate it! I think there an opportunity here for an open community discussion around the use of YAML as the 'source code format' - possibly from the perspective of:
1. **Diff/Merge** - we obviously want it easily diff/merge in source control and ALM - would alternatives (e.g. JSON/XML/Other) make it easier or harder to diff/merge?
2.  **Authoring Experience** - if it is intended that this 'language' be used to write apps 'by hand' - then is the format going to support that and be easy to learn? Should there be an editor that understands the format and can provide IntelliSense?

A number of starting points to discuss spring to mind:

1. **Verbosity** - Some formats like XML are very verbose - JSON can also be quite verbose especially where you have string escaping inside string escaping. 
*E.g Consider the following  YAML*
```
imgControl As image:
    Image: =If(varCurrentMode=ThisItem.ani,ThisItem.imgSel,ThisItem.imgOff)
    ZIndex: =1
    Width: =82
    Height: =151
    ImagePosition: =ImagePosition.Fill
    OnSelect: |
        // explode all lemmings
        UpdateIf(colLemmings,ani<>"" && Not(ani in ["explode","explode_done","splut_done"]),{
            countdown:10
        });
```
In JSON the OnSelect value would look far more verbose with lots of escaping - this does make it harder to read.
```
{
  "imgControl As image": {
    "Image": "If(varCurrentMode=ThisItem.ani,ThisItem.imgSel,ThisItem.imgOff)",
    "ZIndex": 1,
    "Width": 82,
    "Height": 151,
    "ImagePosition": "ImagePosition.Fill",
    "OnSelect": "// explode all lemmings\n                    UpdateIf(colLemmings,ani<>\"\" && Not(ani in [\"explode\",\"explode_done\",\"splut_done\"]),{\n                        countdown:10\n                    });"
  }
}
```
> What do folks think - is there a better option to YAML from this perspective?

2. **Indentation/Escaping/New Lines** - If this format is intended to be written by hand then how hard will the format be for people to learn? Especially where there are multi-line expressions that all need to be indented the same as the parent containing YAML element
*E.g. Consider the following that will pack:*
```
        imgControl As image:
            Image: =If(varCurrentMode=ThisItem.ani,ThisItem.imgSel,ThisItem.imgOff)
            ZIndex: =1
            Width: =82
            Height: =151
            ImagePosition: =ImagePosition.Fill
            OnSelect: |
                =Set(varCurrentMode,ThisItem.ani);
                If(ThisItem.ani="nuke",
```
but the following will throw the error `Can't have empty multiline expressions.` or `Missing ':'. If this is a multiline property, use | at ...` (because the =Set is not indented correctly).
```
        imgControl As image:
            Image: =If(varCurrentMode=ThisItem.ani,ThisItem.imgSel,ThisItem.imgOff)
            ZIndex: =1
            Width: =82
            Height: =151
            ImagePosition: =ImagePosition.Fill
            OnSelect: |
=Set(varCurrentMode,ThisItem.ani);
            If(ThisItem.ani="nuke",
```
This will make it difficult to copy/paste code between expressions - without some IDE tooling.

> What do folks think about the alternatives to YAML from the perspective of indentation/escaping/newlines?

3.  *Consistency* - How important do folks value this? Currently, some of the elements are unpacked as json and some as yaml - should there be a sinle format for all elements apart from resources such as images?
For perspective - for a large app, the majority of the files in the unpacked files are JSON and XML (Data sources, Resource metadata, etc.)
 
> What do folks think - is consistency important?

**Comment by MikeStall** [comment]
"Indentation/Escaping/New Lines" is a huge issue here. It was a top priority that the format here was human readable and editable, and that means users could copy and paste directly between various usages: studio's formula bar, the text format in VS code, documentation, and online answers they may post in forums. 

For comparison, consider this formula for a Text property:
```
  "Hello" & 
      " World"
```

Here's how it looks in various encodings: 
JSon:
```
"Text" : "\"Hello\" & \n  \" World\""
```

XML:
```
<text>&quot;Hello&quot; &amp; 
   &quot;World&quot;
</text>
```

YAML:
```
Text: |
   "Hello" & 
     "World"
```

Yaml is the only one that allow direct copy & paste between studio's formula bar (although you need to left-indent) . Yaml trades escape characters (\", \n, etc) for significant whitespace (left indentation). Practically, this meant that Json and XML were non-starters, and our decision came down to Yaml vs. New format (where that new format would still have significant whitespace, but be highly tailored to the existing PowerApps formula grammar).

**Comment by Kato88** [comment]
What about XML and the usage of cdata? Wouldn't it make your example to

`<text>
<![CDATA["Hello" &
"World"]]>
</text>`

I don't know if this could work but would reduce the amount of chars that needed to be escaped.
Although I don't really see a big benefit in choosing XML over yaml. It still had the same issues with length and size.

**Comment by smfpride** [comment]
@MikeStall I comprehend what you are saying, but I'm not sure that I understand the reasoning if that makes any sense. 

I think it would help to understand who the target audience is for the hand writing/editing of the yaml is. Reason being, I feel that the subset of yaml, paired with the precise formatting required to make it work, makes the target audience not citizen developers, but "pro" developers. Following this line of thinking, I would imagine that a pro developer would have no issue sifting through the escaped characters in an xml or json format. I can't see any citizen developers that I've worked with performing any hand modifications to yaml, json, or xml. 

Additionally, the studio formula bar can be modified in such a way as to parse the escaped line breaks and formatting, we have done similar in other browser based wysiwyg editors when pasting content from Word or Outlook, amongst other things.

**Comment by MikeStall** [comment]
@Kato88 - CData is still very verbose, especially because this is an issue for *every* property. There can be hundreds of properties in a file (see Scott's original example at the top of the thread).  

And _it's often not obvious realizing when you need to escape_ (perhaps one troublesome character in a long string).  
For example, consider changing an 'equals' to 'less than or equals'. In the yaml approach, the distinction doesn't matter, and you can just copy & paste the formula directly from studio's formula bar to the file.  

```
IsVisible1: = A=B
IsVisible2: = A<=B
```

But in xml, 
```
<IsVisible1>A=B</IsVisible1>
<IsVisible2>A&lt;=B</IsVisible2>
```
If we start adding CDAtas to the properties, it gets really verbose.

**Comment by MikeStall** [comment]
@smfpride  - 
We would like to enable humans editing the source files (pa.yaml) in VS Code and other IDEs, particularly for cases where the focus is on formulas instead of UI layout. Screens that are mostly UI layout will still be best to do in Studio. 

We think a pro-developer could sift through, manage these different contexts, and escape as appropriate, but that it's very easy to get wrong and would be a constant source of unnecessary grief. And often missing escaping may not produce a syntax error but actually silently change the semantics.   Plus it can end up creating 2 flavors of the language (the escaped version that pro-devs use with source files; and the non-escaped version used by citizen devs in formula bar).  That could create barriers in collaboration, especially in places like sharing snippets in documentation or forum answers. 

Basically, we want the same formula text that can be copied and pasted between these places:

- Formula bar in studio 
- Documentation examples
- Stackoverflow / Forum responses
- source files in VSCode.  

We'd love to hear more from customers on this topic.

**Comment by amervitz** [comment]
The desire of combining data structures and  code in one file is what I see as being the main challenge to finding a file format that can store both, without needing any escaping or other formatting restrictions. Why not keep the different types of data in their own respective files that then allow the information in each to be represented in their most natural format?

**Comment by MikeStall** [comment]
@amervitz - that's exactly what we do. For example, Connector information and API definitions are still in .json.  But formulas are in .pa.yaml.

**Comment by amervitz** [comment]
@MikeStall the expressions in the .pa.yaml cannot be expressed in their raw format though, and the file contains properties with and without expressions. I'm suggesting the expressions should not be in the yaml file but externally in a separate file.

**Comment by scottdurow** [comment]
@amervitz I initially thought the same, but having used pasopa for a while on some very large apps - it feels good having the expressions in the yaml so you can see all the control properties in one place - rather than continually swapping between files and formats.

I think the exception would be for supporting common functions - where the function code could be stored in a different file and then referenced by the pa.yaml

**Comment by MikeStall** [comment]
@amervitz - that yaml file **is** the "externally separate file"!   

Every property is an expression (although most properties are trivial expressions such as a single number).   So if we put each formula in its own file, that could result in 10000s of files for a canvas app. We need some format allows multiple properties to be placed in a single file, without escaping. That's yaml (or could have been a new format).  

See example here: https://github.com/microsoft/PowerApps-Language-Tooling/blob/master/src/PAModelTests/SrcBaseline/Weather_Screen1.pa.yaml

**Comment by amervitz** [comment]
@scottdurow I can't argue with the productivity benefits of having the expressions in the same file.

Are "common functions" a new concept or already possible? If new, I've asked for this and would like to see msft design this. If existing, if like to see an example.

**Comment by Kato88** [comment]
The question is whether it is more problematic to have thousands of lines of YAML files with countless indentations or to have thousands of smaller files with fewer indentations. Sounds to me like you should go with the lesser of two evils, and I see YAML as the best option at the moment.

**Comment by mariusagur** [comment]
I think indentation gets exponentially worse as the number of lines grows, so for me it's just two different evils.
With the risk of saying something stupid, I would prefer there was a superset which transpiles to whatever you end up with. 
Something that would enable you to create sections and link to files/other sections as you wish. This would also be great for snippets and creating a in-house library with all the benefits that gives you.
I may be mistaken here, but I think indentation is a big pain for accessibility/universal design, so a superset, if at all possible, could potentially enable more people to use it

**Comment by Kato88** [comment]
> I think indentation gets exponentially worse as the number of lines grows

I think there is better tooling for indentations than for long files. And by todays standard you probably have a wider than longer screen. In the end it comes down to preferences?

Maybe another very bold (or stupid) idea, but could you do subfiles for pa.yaml? Like one file per container?

`|
|- Screen1.pa.yaml
|--- MainContainer.pa.yaml
|----- LeftContainer.pa.yaml
|----- RightContainer.pa.yaml`

In Visual Studio you can do that (I don't know how it is called but you can put files under a specific file. t4 Templates are a prime example)

**Comment by MikeStall** [comment]
@Kato88 -  The challenge is that sometimes people want them all in a single screen file (often if it's small) and other times they want them in multiple files.  In C# and other languages, the compiler lets you organize the source files however you'd like. 

However, With pa.yaml, we need to pack **and unpack**, and to avoid "noisy diffs", we need the unpacked sources to match exactly with the packed sources. 
In contrast, other languages like C# don't have this problem. C# only "packs" (ie, compiles).   (While you can decompile C#  there's no expectation the decompiled code match line-for-line with the original sources. ) 

If we allowed more arbitrary file layouts, we'd need to track the file mapping in the msapp so that we could faithfully decompile. And we'd need it to be smart enough to deal with edits from studio.  For example, if you added a new control in Screen1.pa.yaml, which sub file should it go into. 

If you'd like to explore this "file granularity" problem further, can you file a dedicated issue for just that and we can discuss further? (It's worth noting it's also orthogonal to whether we use yaml or some other format).
Thanks!

**Comment by amervitz** [comment]
@MikeStall what's the roadmap for common functions? If they were in a separate file and the yaml file were just a call to the common function by name with parameters, the file containing the common functions could be a pure source code file without any yaml restrictions.

**Comment by MikeStall** [comment]
@amervitz  - 
One challenge here is that we want the source format to be 100% faithful to the server format. This means that the file must round-trip and the server must understand all the concepts present in the source files. This means that we avoid client-side transforms like preprocessors and client-side build steps.  So while somebody could emulate code-reuse here by having a client-side script that copies the code, the server wouldn't capture the intent and once you unpacked the resulting msapp again, it wouldn't just be repeated code. 

That said, Functions are definitely something we'd like to do... stay tuned to https://powerapps.microsoft.com/en-us/blog/ for announcements.

**Comment by amervitz** [comment]
If no further details of functions can be provided outside of announcements, I'm on board with using YAML for properties if:
- there will definitely be future support for said functions
- the functions will allow the raw source code for multi-line expressions to be stored in separate file(s) and not require any escaping

**Comment by rappen** [comment]
[shameless plug]

We're having an interesting discussion with @MikeStall on this topic on the latest XrmToolCast episode.
https://go.xrmtoolcast.com/pasopa

**Comment by landonarogers** [comment]
Thank you all for your input on this issue! We've collected your responses and we appreciate the time you've spent on discussing how to improve this repo!
